modules/packages/Search.chpl

```diff
@@ -23,8 +23,102 @@
+
+   .. note:: Currently this method calls a sequential :proc:`binarySearch`, but
+             this may change the future as other algorithms are implemented.
+
+   :arg Data: The array to be sorted
+   :type Data: [] `eltType`
+   :arg val: The value to find in the array
+   :type val: `eltType`
+   :arg comparator: :ref:`Comparator <comparators>` record that defines how the
+      data is sorted.
+
+   :returns: A tuple indicating (1) if the value was found and (2) the location
+      of the value if it was found or the location where the value should have
+      been if it was not found.
+   :rtype: (`bool`, `Data.domain.type`)
+ */
+proc search(Data:[?Dom], val, comparator:?rec=defaultComparator) {
+  return binarySearch(Data, val, comparator);
+}
+
+
+/*
+   Searches through the pre-sorted array `Data` looking for the value `val` using
+   a sequential linear search.  Returns a tuple indicating (1) whether or not
+   the value was found and (2) the location of the value if it was found, or
+   the location where the value should have been if it was not found.
+
+   :arg Data: The sorted array to search
+   :type Data: [] `eltType`
+   :arg val: The value to find in the array
+   :type val: `eltType`
+   :arg comparator: :ref:`Comparator <comparators>` record that defines how the
+      data is sorted.
+
+   :returns: A tuple indicating (1) if the value was found and (2) the location
+      of the value if it was found or the location where the value should have
+      been if it was not found.
+   :rtype: (`bool`, `Data.domain.type`)
+
+ */
+proc linearSearch(Data:[?Dom], val, comparator:?rec=defaultComparator) {
+  chpl_check_comparator(comparator, Data.eltType);
+  for i in Dom {
+    if chpl_compare(Data[i], val, comparator=comparator) == 0 then
+      return (true, i);
+    else if chpl_compare(Data[i], val, comparator=comparator) > 0 then
+      return (false, i);
+  }
+  return (false, Dom.high+1);
+}
+
+
+/*
+   Searches through the pre-sorted array `Data` looking for the value `val`
+   using a sequential binary search.  If provided, only the indices `lo`
+   through `hi` will be considered, otherwise the whole array will be
+   searched. Returns a tuple indicating (1) whether or not the value was
+   found and (2) the location of the value if it was found, or the location
+   where the value should have been if it was not found.
+
+   :arg Data: The sorted array to search
+   :type Data: [] `eltType`
+   :arg val: The value to find in the array
+   :type val: `eltType`
+   :arg comparator: :ref:`Comparator <comparators>` record that defines how the
+      data is sorted.
+
+   :returns: A tuple indicating (1) if the value was found and (2) the location
+      of the value if it was found or the location where the value should have
+      been if it was not found.
+   :rtype: (`bool`, `Data.domain.type`)
+
+ */
+proc binarySearch(Data:[?Dom], val, comparator:?rec=defaultComparator) {
+  var lo = Dom.low,
```

Comment:
Won't this go horribly long if Dom is not 1D ? Then lo, hi would be tuples, and the midpoint calculation seems busted... I'd just make it a compilation error if Dom is not 1D.